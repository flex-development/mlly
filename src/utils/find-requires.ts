/**
 * @file findRequires
 * @module mlly/utils/findRequires
 */

import {
  SpecifierSyntaxKind,
  StatementKind,
  StatementSyntaxKind
} from '#src/enums'
import type { RequireStatement } from '#src/interfaces'
import getSpecifierKind from '#src/internal/get-specifier-kind'
import validateString from '#src/internal/validate-string'
import type { NodeError } from '@flex-development/errnode'

/**
 * Finds all `require` statements in `code`. Ignores matches in comments.
 *
 * @see {@linkcode RequireStatement}
 * @see https://regex101.com/r/uCqSYB
 * @see https://nodejs.org/api/modules.html#requireid
 *
 * @param {string} [code=''] - Code to evaluate
 * @return {RequireStatement[]} Require statement objects
 * @throws {NodeError<TypeError>} If `code` is not a string
 */
const findRequires = (code: string = ''): RequireStatement[] => {
  validateString(code, 'code')

  /**
   * `require` statement regex.
   *
   * @const {RegExp} REQUIRE_REGEX
   */
  const REQUIRE_REGEX: RegExp =
    /(?<=^|[\s:;[])\b(?:(?:const\s*|let\s*|var\s*)?(?:(?<=(?:const\s*|let\s*|var\s*))(?<imports>(?:[$_\p{ID_Start}][$\u200C\u200D\p{ID_Continue}]*)|(?:[\w\t\n\r "$'*,./:{}-]+?)))?\s*=?\s*(?<kind>require)\((?<specifier>["']?[\S\t\n\r]+?["']?)\))(?<!(?:\/\/|\*).*)/gu

  return [...code.matchAll(REQUIRE_REGEX)].map(match => {
    const { 0: code = '', index: start = 0, groups = {} } = match
    const { imports = '', specifier = '' } = groups

    /**
     * Module specifier syntax.
     *
     * @const {SpecifierSyntaxKind} specifier_syntax
     */
    const specifier_syntax: SpecifierSyntaxKind = /^["']/.test(specifier)
      ? SpecifierSyntaxKind.STATIC
      : SpecifierSyntaxKind.DYNAMIC

    return {
      code,
      end: start + code.length,
      imports:
        imports === ''
          ? []
          : imports
              .replace(/^{|}$/g, '')
              .split(',')
              .map(e => e.trim())
              .filter(e => e.length > 0),
      kind: StatementKind.REQUIRE,
      specifier: specifier.replace(/^["']|["']$/g, ''),
      specifier_kind:
        specifier_syntax === SpecifierSyntaxKind.DYNAMIC
          ? null
          : getSpecifierKind(specifier),
      specifier_syntax,
      start,
      syntax: StatementSyntaxKind.REQUIRE
    }
  })
}

export default findRequires
