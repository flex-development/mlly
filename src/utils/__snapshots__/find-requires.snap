// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:utils/findRequires > should return RequireStatement object array 1`] = `
[
  {
    "code": "require('./side-effect.cjs')",
    "end": 28,
    "imports": [],
    "kind": "require",
    "specifier": "./side-effect.cjs",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 0,
    "syntax": "require",
  },
  {
    "code": "const mod=require('./lib')",
    "end": 55,
    "imports": [
      "mod",
    ],
    "kind": "require",
    "specifier": "./lib",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 29,
    "syntax": "require",
  },
  {
    "code": "const{createPathMatch}=require('tsconfig-paths')",
    "end": 105,
    "imports": [
      "createPathMatch",
    ],
    "kind": "require",
    "specifier": "tsconfig-paths",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 57,
    "syntax": "require",
  },
  {
    "code": "const mod = require('lib')",
    "end": 133,
    "imports": [
      "mod",
    ],
    "kind": "require",
    "specifier": "lib",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 107,
    "syntax": "require",
  },
  {
    "code": "require(id)",
    "end": 146,
    "imports": [],
    "kind": "require",
    "specifier": "id",
    "specifier_kind": null,
    "specifier_syntax": "dynamic",
    "start": 135,
    "syntax": "require",
  },
  {
    "code": "require('#src/utils/resolve-aliases')",
    "end": 184,
    "imports": [],
    "kind": "require",
    "specifier": "#src/utils/resolve-aliases",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 147,
    "syntax": "require",
  },
  {
    "code": "require("./arr")",
    "end": 214,
    "imports": [],
    "kind": "require",
    "specifier": "./arr",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 198,
    "syntax": "require",
  },
  {
    "code": "require("./config")",
    "end": 258,
    "imports": [],
    "kind": "require",
    "specifier": "./config",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 239,
    "syntax": "require",
  },
  {
    "code": "require("./defaults")",
    "end": 314,
    "imports": [],
    "kind": "require",
    "specifier": "./defaults",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 293,
    "syntax": "require",
  },
  {
    "code": "require("./conf")",
    "end": 332,
    "imports": [],
    "kind": "require",
    "specifier": "./conf",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 315,
    "syntax": "require",
  },
  {
    "code": "require("./data")",
    "end": 374,
    "imports": [],
    "kind": "require",
    "specifier": "./data",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 357,
    "syntax": "require",
  },
  {
    "code": "require("./parser-options")",
    "end": 428,
    "imports": [],
    "kind": "require",
    "specifier": "./parser-options",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 401,
    "syntax": "require",
  },
  {
    "code": "const {
  default: myDefault,
  addFive,
	addFour,
	addThree,
	addTwo,
	squareFive,
	squareFour,
	squareThree,
	squareTwo
} = require('./lib')",
    "end": 572,
    "imports": [
      "default: myDefault",
      "addFive",
      "addFour",
      "addThree",
      "addTwo",
      "squareFive",
      "squareFour",
      "squareThree",
      "squareTwo",
    ],
    "kind": "require",
    "specifier": "./lib",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 430,
    "syntax": "require",
  },
]
`;
